onnx/defs/schema.h

```diff
@@ -1423,7 +1448,12 @@ class OpSchemaRegistry final : public ISchemaRegistry {
   }
 };
 
-void RegisterSchema(OpSchema schema, int opset_version_to_load = 0, bool fail_duplicate_schema = true);
+void RegisterSchema(
+    OpSchema schema,
```

Comment:
```suggestion
    OpSchema& schema,
```

can this take a reference instead?

Reply:
Can https://github.com/onnx/onnx/blob/48501e0ff59843a33f89e028cbcbf080edd0c1df/onnx/defs/schema.h#L1257C24-L1257C30 still be moved, or does it need to be copied? I am not sure.

Reply:
Using non-const reference can be risky, particularly when we plan to move the referenced object using std::move.

Current register workflow:

 - Buildin schema: `var: T&&` -> `T&&`[1] -> `T`[2] -> `T&`[3] -> `T&`[4] -> `std::move`
 - Python custom schema: `var: T` -> `T`[2] -> `T&`[4] -> `std::move`
 - Cpp custom schema: `var: T` -> `T&`[4] -> `std::move`

We can retain the use of move, and changing non-reference argument in final function[4] to avoid the risks associated with non-const reference. But it might introduce additional overhead. So I haven't pushed this commit yet.
After update, the chain will be:

 - Buildin schema: `var: T&&` -> `T&&`[1] -> `const T&`[2] -> `const T&`[3] -> `T`[4] -> `std::move`
 - Python custom schema: `var: T` -> `const T&`[2] -> `T`[4] -> `std::move`
 - Cpp custom schema: `var: T` -> `T`[4] -> `std::move`

[1] `RegisterOpSetSchema<T>`
[2] `RegisterSchema`
[3] `OpSchemaRegisterOnce`
[4] `OpSchemaRegisterImpl`

Reply:
Thanks for the clarification. I think the change makes sense to avoid the ambiguities associated with the reference in [4]. @edgchen1 @gramalingam could you chime in / do you have any recommendations? Thanks!

Reply:
For Built-in schemas, we can move everything. Python custom schemas need to be owned by Python so there needs to be at least one copies. I am not too sure about Cpp custom schemas.

Reply:
Should we care about the performance of this function? For this interface issue, I prefer code standardization.

Reply:
We do care about performance (both compute and memory overhead) but I think code readability outweighs the cost in this case, as the duplicated opschemas are eventually released and do not take much memory in the first place. Please correct me if I am missing anything.

Reply:
I have push a commit `Safe implementation for register` to implemented the above changes. We can modify and reverse at any time.

Reply:
Hi, I am not convinced about the tradeoff. First, my impression is that this code doesn't have to change at all for the real purpose of this PR (which just requires a new call to RegisterSchema for a schema coming from Python), or am I mistaken? It seems simpler to focus on just adding the new functionality, without touching what is, admittedly, some complex piece of code (from ownership perspective).

If it was just my execution time vs. my dev time, I would gladly make this tradeoff. But it seems to me the real tradeoff is between increased execution time for potentially millions of executions of this code every day for what gain? It seems to me readability can be achieved in other ways without giving up performance (eg. just clearly documenting what is happening).

Anyway, that is just my 2c worth.

Reply:
I understand that the added copy only happens at registration time. So this change adds one copy for each schema at registration. I don’t think this is going to be the bottleneck or create meaningful impact? On the other hand, moving a reference can be error prone because the signature doesn’t assume ownership. Please feel free to correct me if I am missing anything.

Reply:
Why can't the move happen up the stack earlier to clarify the ownership transfer?

Reply:
Specifically, why can't [2], [3], [4] all take `T&&` parameter?

Reply:
Thanks for your suggestion. Revert the commit, and using `T&&` throughout the pipeline.